File: MinecraftMediaLibrary/src/main/java/com/github/pulsebeat02/test/YoutubeResourcepackTest.java
package com.github.pulsebeat02.test;

import com.github.pulsebeat02.MinecraftMediaLibrary;
import com.github.pulsebeat02.concurrent.AsyncVideoExtraction;
import com.github.pulsebeat02.extractor.YoutubeExtraction;
import com.github.pulsebeat02.image.MapImage;
import com.github.pulsebeat02.resourcepack.ResourcepackWrapper;
import com.github.pulsebeat02.resourcepack.hosting.HttpDaemonProvider;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class YoutubeResourcepackTest extends JavaPlugin {

    private MinecraftMediaLibrary library;

    @Override
    public void onEnable() {
        library = new MinecraftMediaLibrary(this, getDataFolder().getPath(), true);
    }

    public String getResourcepackUrlYoutube(@NotNull final String youtubeUrl, @NotNull final String directory, final int port) {

        final YoutubeExtraction extraction = new YoutubeExtraction(youtubeUrl, directory) {
            @Override
            public void onVideoDownload() {
                System.out.println("Video is Downloading!");
            }

            @Override
            public void onAudioExtraction() {
                System.out.println("Audio is being extracted from Video!");
            }
        };
        final ExecutorService executor = Executors.newCachedThreadPool();
        CompletableFuture.runAsync(() -> new AsyncVideoExtraction(extraction).extractAudio(), executor);
        CompletableFuture.runAsync(() -> new AsyncVideoExtraction(extraction).downloadVideo(), executor);

        final ResourcepackWrapper wrapper = new ResourcepackWrapper.Builder()
                .setAudio(extraction.getAudio())
                .setDescription("Youtube Video: " + extraction.getVideoTitle())
                .setPath(directory)
                .setPackFormat(6)
                .createResourcepackHostingProvider(library);
        wrapper.buildResourcePack();

        final HttpDaemonProvider hosting = new HttpDaemonProvider(directory, port);
        hosting.startServer();

        return hosting.generateUrl(Paths.get(directory));

    }

    public void displayImage(final int map, @NotNull final File image) throws IOException {

        final BufferedImage bi = ImageIO.read(image);

        final MapImage imageMap = new MapImage.Builder()
                .setMap(map)
                .setWidth(bi.getWidth())
                .setHeight(bi.getHeight())
                .createImageMap(library);

        imageMap.drawImage();

    }

}

File: api/src/test/java/com/github/pulsebeat02/minecraftmedialibrary/test/video/YoutubeResourcepackTest.kt
/*............................................................................................
 . Copyright © 2021 Brandon Li                                                               .
 .                                                                                           .
 . Permission is hereby granted, free of charge, to any person obtaining a copy of this      .
 . software and associated documentation files (the “Software”), to deal in the Software     .
 . without restriction, including without limitation the rights to use, copy, modify, merge, .
 . publish, distribute, sublicense, and/or sell copies of the Software, and to permit        .
 . persons to whom the Software is furnished to do so, subject to the following conditions:  .
 .                                                                                           .
 . The above copyright notice and this permission notice shall be included in all copies     .
 . or substantial portions of the Software.                                                  .
 .                                                                                           .
 . THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND,                           .
 .  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF                       .
 .   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                                   .
 .   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS                     .
 .   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN                      .
 .   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN                       .
 .   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE                        .
 .   SOFTWARE.                                                                               .
 ............................................................................................*/
package com.github.pulsebeat02.minecraftmedialibrary.test.video

import com.github.pulsebeat02.minecraftmedialibrary.MediaLibrary
import com.github.pulsebeat02.minecraftmedialibrary.concurrent.AsyncVideoExtraction
import com.github.pulsebeat02.minecraftmedialibrary.extractor.ExtractionSetting
import com.github.pulsebeat02.minecraftmedialibrary.extractor.YoutubeExtraction
import com.github.pulsebeat02.minecraftmedialibrary.image.basic.MinecraftStaticImage
import com.github.pulsebeat02.minecraftmedialibrary.resourcepack.ResourcepackWrapper
import com.github.pulsebeat02.minecraftmedialibrary.resourcepack.hosting.HttpDaemonProvider
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import javax.imageio.ImageIO

class YoutubeResourcepackTest : JavaPlugin() {

    private var library: MediaLibrary? = null

    override fun onEnable() {
        library = MediaLibrary(this)
    }

    fun getResourcepackUrlYoutube(youtubeUrl: String, directory: String, port: Int): String {
        val extraction: YoutubeExtraction = object : YoutubeExtraction(
            youtubeUrl, directory, ExtractionSetting.builder().build()
        ) {
            override fun onVideoDownload() {
                println("Video is Downloading!")
            }

            override fun onAudioExtraction() {
                println("Audio is being extracted from Video!")
            }
        }
        val executor = Executors.newCachedThreadPool()
        CompletableFuture.runAsync({ AsyncVideoExtraction(extraction).extractAudio() }, executor)
        CompletableFuture.runAsync(
            { AsyncVideoExtraction(extraction).downloadVideo() }, executor
        )
        val wrapper = ResourcepackWrapper.builder()
            .setAudio(extraction.audio)
            .setDescription("Youtube Video: " + extraction.videoTitle)
            .setPath(directory)
            .setPackFormat(6)
            .build(library)
        wrapper.buildResourcePack()
        val hosting = HttpDaemonProvider(directory, port)
        hosting.startServer()
        return hosting.generateUrl(Paths.get(directory))
    }

    @Throws(IOException::class)
    fun displayImage(map: Int, image: File) {
        val bi = ImageIO.read(image)
        val imageMap = MinecraftStaticImage.builder()
            .setMap(map)
            .setWidth(bi.width)
            .setHeight(bi.height)
            .build(library)
        imageMap.drawImage()
    }
}